core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestRestoreSnapshotAction.java

```diff
@@ -49,7 +50,11 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         RestoreSnapshotRequest restoreSnapshotRequest = restoreSnapshotRequest(request.param("repository"), request.param("snapshot"));
         restoreSnapshotRequest.masterNodeTimeout(request.paramAsTime("master_timeout", restoreSnapshotRequest.masterNodeTimeout()));
         restoreSnapshotRequest.waitForCompletion(request.paramAsBoolean("wait_for_completion", false));
-        restoreSnapshotRequest.source(request.content().utf8ToString());
+        if (request.hasContent()) {
```

Comment:
maybe for usages like this one we can add a helper method on `RestRequest`? Like this one:

```java
public final void applySource(Consumer<Map<String, Object>> target) throws IOException {
    if (hasContent()) {
        try (XContentParser parser = contentParser()) {
            target.accept(parser.mapOrdered());
        }
   }
}
```

Then this snippet can be replaced with:

```java
request.applySource(restoreSnapshotRequest::source);
```

and we could have a similar helper that passes a `XContentParser` to an action request?